gcd/src/main/java/org/spine3/server/storage/datastore/DsFilters.java

```diff
@@ -0,0 +1,472 @@
+                                                                .name());
+            }
+        }
+
+        /**
+         * Traverses the tree of the {@linkplain ColumnFilterNode column filters} starting from
+         * the current node, passing all the found paths from the tree top to a leaf to the given
+         * result processor.
+         *
+         * @param processor the result processor
+         */
+        @SuppressWarnings("MethodWithMultipleLoops")
+            // To make the traversal algorithm more obvious.
+        private void traverse(ConjunctionProcessor processor) {
+            final Queue<ColumnFilterNode> nodes = new LinkedList<>();
+            final Queue<Queue<ColumnFilterNode>> paths = new LinkedList<>();
+            nodes.offer(this);
+            // Initial path is intentionally left empty.
+            paths.offer(new LinkedList<ColumnFilterNode>());
+
+            while (!nodes.isEmpty()) {
+                final ColumnFilterNode node = nodes.poll();
+                final Queue<ColumnFilterNode> path = paths.poll();
+
+                if (node.isLeaf()) {
+                    processor.process(path);
+                } else {
+                    for (ColumnFilterNode child : node.subtrees) {
+                        final Queue<ColumnFilterNode> childPath = newPath(path);
+                        childPath.offer(child);
+                        nodes.offer(child);
+                        paths.offer(childPath);
+                    }
+                }
+            }
+        }
+
+        private boolean isLeaf() {
+            return subtrees.isEmpty();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            ColumnFilterNode that = (ColumnFilterNode) o;
+            return Objects.equal(getColumn(), that.getColumn()) &&
+                    Objects.equal(getColumnFilter(), that.getColumnFilter());
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(getColumn(), getColumnFilter());
+        }
+    }
+
+    /**
+     * A specific type of a {@code ColumnFilterNode} tree node representing the tree head.
+     */
+    private static final class ColumnFilterTreeHead extends ColumnFilterNode {
+
+        private static ColumnFilterNode instance() {
+            return new ColumnFilterTreeHead();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            return o == this;
+        }
+
+        @Override
+        public int hashCode() {
+            return 0;
+        }
+    }
+
+    /**
+     * A functional interface defining the operation of precessing a conjunction group found by
+     * the {@linkplain #fromParams parenthesis opening algorithm}.
+     */
+    private interface ConjunctionProcessor {
```

Comment:
This doesn't look a proper usage of `Function` to me.
Moreover, the custom type provides a description of the usage of the type instances and `Function<Node, Void>` says nothing about both the function itself and the parameter.